cpp/src/arrow/util/iterator.h

```diff
@@ -249,6 +262,32 @@ Iterator<T> MakeVectorIterator(std::vector<T> v) {
   return Iterator<T>(VectorIterator<T>(std::move(v)));
 }
 
+/// \brief Simple iterator which yields the elements of a std::vector<T> as optional<T>.
+/// This is provided to support T where IterationTraits<T>::End is not specialized
+template <typename T>
+class VectorOptionalIterator {
+ public:
+  explicit VectorOptionalIterator(std::vector<T> v) : elements_(std::move(v)) {}
+
+  Status Next(util::optional<T>* out) {
+    if (i_ == elements_.size()) {
```

Comment:
Since the operands to the equivalent ternary operator would have different type (`nullopt_t` vs `T`) they'd need to be wrapped in `optional<T>`. I prefer this form